reference/matrix/ell_kernels.cpp

```diff
@@ -208,6 +208,30 @@ GKO_INSTANTIATE_FOR_EACH_VALUE_AND_INDEX_TYPE(
     GKO_DECLARE_ELL_CALCULATE_NONZEROS_PER_ROW_KERNEL);
 
 
+template <typename ValueType, typename IndexType>
+void extract_diagonal(std::shared_ptr<const ReferenceExecutor> exec,
+                      const matrix::Ell<ValueType, IndexType> *orig,
+                      matrix::Dense<ValueType> *diag)
+{
+    const auto col_idxs = orig->get_const_col_idxs();
+    const auto values = orig->get_const_values();
+    const auto diag_size = diag->get_size()[0];
+    const auto max_nnz_per_row = orig->get_num_stored_elements_per_row();
+
+    for (size_type row = 0; row < diag_size; row++) {
+        for (size_type i = 0; i < max_nnz_per_row; i++) {
+            if (orig->col_at(row, i) == row) {
```

Comment:
check the value is equal to zero or not?

Reply:
I put the nonzero check into the cuda and hip implementation to check if it's an actual entry or en empty one. So with the break here after finding the diagonal entry in the row, this is not necessary